File: src/main/java/org/apache/prepbuddy/datacleansers/imputation/ApproxMeanSubstitution.java
package org.apache.prepbuddy.datacleansers.imputation;

import org.apache.prepbuddy.rdds.TransformableRDD;
import org.apache.prepbuddy.utils.RowRecord;
import org.apache.spark.api.java.JavaDoubleRDD;

/**
 * A imputation strategy that uses meanApprox method of JavaDoubleRDD to impute the missing
 * value by approx mean.
 */
public class ApproxMeanSubstitution implements ImputationStrategy {
    private Double approximateMean;

    @Override
    public void prepareSubstitute(TransformableRDD rdd, int missingDataColumn) {
        int timeout = 20000;
        JavaDoubleRDD javaDoubleRDD = rdd.toDoubleRDD(missingDataColumn);
        this.approximateMean = javaDoubleRDD.meanApprox(timeout).getFinalValue().mean();
    }

    @Override
    public String handleMissingData(RowRecord record) {
        return approximateMean.toString();
    }
}

File: src/main/scala/org/apache/datacommons/prepbuddy/imputations/ApproxMeanSubstitution.scala
package org.apache.datacommons.prepbuddy.imputations

import org.apache.datacommons.prepbuddy.rdds.TransformableRDD
import org.apache.datacommons.prepbuddy.utils.RowRecord

/**
  * An imputation strategy that imputes the missing values by an approx mean of the values in data set.
  * This implementation is only for imputing numeric columns.
  * Recommended when imputing on large data set.
  */
class ApproxMeanSubstitution() extends ImputationStrategy {
    private var approxMean: Double = 0

    override def prepareSubstitute(rdd: TransformableRDD, missingDataColumn: Int): Unit = {
        val timeOut: Int = 20000
        approxMean = rdd.toDoubleRDD(missingDataColumn).meanApprox(timeOut).getFinalValue().mean
    }

    override def handleMissingData(record: RowRecord): String = approxMean.toString
}
